src/interactive.c

```diff
@@ -54,29 +54,23 @@ interactive_begin(struct view *view, enum input_mode mode, uint32_t edges)
 	if (view->maximized || view->tiled) {
 		if (mode == LAB_INPUT_STATE_MOVE) {
 			/* Exit maximized or tiled mode */
-			int new_x = max_move_scale(view->server->seat.cursor->x,
-				view->x, view->w, view->natural_geometry.width);
-			int new_y = max_move_scale(view->server->seat.cursor->y,
-				view->y, view->h, view->natural_geometry.height);
-			view->natural_geometry.x = new_x;
-			view->natural_geometry.y = new_y;
-			if (view->maximized) {
-				view_maximize(view, false);
-			}
-			if (view->tiled) {
-				view_move_resize(view, view->natural_geometry);
-			}
+			struct wlr_box geometry = view->natural_geometry;
+			geometry.x = max_move_scale(seat->cursor->x, view->x,
+				view->w, geometry.width);
+			geometry.y = max_move_scale(seat->cursor->y, view->y,
+				view->h, geometry.height);
+			view_restore_to(view, geometry);
 			/**
-			 * view_maximize() / view_move_resize() indirectly calls
+			 * view_restore_to() indirectly calls
 			 * view->impl->configure which is async but we are using
 			 * the current values in server->grab_box. We pretend the
 			 * configure already happened by setting them manually.
 			 */
-			server->grab_box.x = new_x;
-			server->grab_box.y = new_y;
-			server->grab_box.width = view->natural_geometry.width;
-			server->grab_box.height = view->natural_geometry.height;
+			server->grab_box = geometry;
 		}
+	} else {
+		/* Store natural geometry at start of move/resize */
+		view_store_natural_geometry(view);
```

Comment:
I think we only have to store the natural geometry when `mode == LAB_INPUT_STATE_MOVE`. Maybe we could switch the `mode == MOVE` check and the `view->maximized || view->tiled check` here so the `mode == MOVE` check is the outer one and the else branch here is for the inner check (`!maximized && !tiled`)

Reply:
Good call. I reorganized `interactive_begin()` so that there is now one `switch(mode)` rather than the various `if` statements.